### PR TITLE
Mesh plane clipping

### PR DESCRIPTION
**Pull request by petersteneteg**
A rebase and refactoring of #711 to support arbitrary meshes (not just simple and basic) also supports lines and points, and multiple index buffers or no index buffer. 

bild

### CODE REVIEW

File: include/inviwo/core/datastructures/geometry/edge.h [hunk 1]
```diff
@@ -35,37 +35,21 @@ namespace inviwo {
 template <typename T>
 class Edge {
```
Comment (ResearchDaniel) [inline include/inviwo/core/datastructures/geometry/edge.h:36]: Could add documentation. I assume v1->v2, e.g., the edge is assumed to start at v1 and end at v2.
Assumes that T is index (UInt32) to a position or the actual positions (vec2, vec3)?

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * Compute barycentric coordinates/weights for
+ * point p (which is inside the polygon) with respect to polygons of vertices (v)
+ * Based on Mean Value Coordinates by Hormann/Floater
+ * following https://www.mn.uio.no/math/english/people/aca/michaelf/papers/barycentric.pdf
+ */
+IVW_MODULE_BASE_API std::vector<float> barycentricInsidePolygon(vec2 v,
+                                                                const std::vector<vec2>& vis);
+
+IVW_MODULE_BASE_API std::optional<glm::u32vec2> sutherlandHodgman(
```
Comment (ResearchDaniel) [inline modules/base/include/modules/base/algorithm/mesh/meshclipping.h:66]: Missing documentation

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    glm::u32vec3 triangle, const Plane& plane, const std::vector<vec3>& positions,
+    std::vector<std::uint32_t>& indices, const InterpolateFunctor& addInterpolatedVertex);
+
+IVW_MODULE_BASE_API void removeDuplicateEdges(std::vector<glm::u32vec2>& cuts,
```
Comment (ResearchDaniel) [inline modules/base/include/modules/base/algorithm/mesh/meshclipping.h:70]: Missing documentation. cuts?

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                                              const std::vector<vec3>& positions, float eps);
+
+IVW_MODULE_BASE_API std::vector<std::vector<std::uint32_t>> gatherLoops(
```
Comment (ResearchDaniel) [inline modules/base/include/modules/base/algorithm/mesh/meshclipping.h:73]: Missing documentation

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+}  // namespace detail
+
+/**
+ * Clip mesh against plane using Sutherland-Hodgman.
+ * If holes should be closed, the input mesh must be manifold.
+ * Vertex attributes that are interpolated and saved: tex coords and color.
+ * Supported mesh types: SimpleMesh and BasicMesh.
+ * Supported draw types: Triangle.
+ * Supported connectivity types: Strip and None.
```
Comment (mskr) [inline modules/base/include/modules/base/algorithm/mesh/meshclipping.h]: should update this

File: modules/base/src/algorithm/mesh/meshclipping.cpp [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+
+    constexpr float relError = 0.000001f;
+    const auto eps =
+        relError * std::accumulate(edges.begin(), edges.end(), 0.0f, [&](float m, glm::u32vec2 e) {
+            return glm::max(m, glm::max(glm::compMax(glm::abs(positions[e[0]])),
+                                        glm::compMax(glm::abs(positions[e[1]]))));
+        });
+
+    removeDuplicateEdges(edges, positions, eps);
+    const auto loops = gatherLoops(edges, positions, eps);
```
Comment (mskr) [inline modules/base/src/algorithm/mesh/meshclipping.cpp:265]: Did you notice any further problems because of the epsilon?

**Review by mskr — APPROVED** [review]
Looks good, except at one point the docs were not updated.

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h (hunk 3 above)
Comment (ResearchDaniel) [reply modules/base/include/modules/base/algorithm/mesh/meshclipping.h:70]: Might want to have a default epsilon (0)?

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h (hunk 4 above)
Comment (ResearchDaniel) [reply modules/base/include/modules/base/algorithm/mesh/meshclipping.h:73]: Might want to have a default epsilon (0)?

**Review by ResearchDaniel — APPROVED** [review]
Looks good. Some clarifications would be good, see comments

File: modules/base/src/algorithm/mesh/meshclipping.cpp (hunk 6 above)
Comment (petersteneteg) [reply modules/base/src/algorithm/mesh/meshclipping.cpp:265]: Not in the meshes I've tried with, but I have not tested with very complex ones...

File: include/inviwo/core/datastructures/geometry/edge.h (hunk 1 above)
Comment (petersteneteg) [reply include/inviwo/core/datastructures/geometry/edge.h:36]: I only cleaned this up a bit, Generally an edge would be undirected I would say.

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h (hunk 2 above)
Comment (petersteneteg) [reply modules/base/include/modules/base/algorithm/mesh/meshclipping.h:66]: Its in a detail namespace, not supposed to be used externally. Needed to expose it for the unit testing...

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h (hunk 3 above)
Comment (petersteneteg) [reply modules/base/include/modules/base/algorithm/mesh/meshclipping.h:70]: no, 0 would not work well

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h (hunk 4 above)
Comment (petersteneteg) [reply modules/base/include/modules/base/algorithm/mesh/meshclipping.h:73]: this is implementation details. only exposed for testing

File: modules/base/include/modules/base/algorithm/mesh/meshclipping.h (hunk 2 above)
Comment (ResearchDaniel) [reply modules/base/include/modules/base/algorithm/mesh/meshclipping.h:66]: Ok

File: modules/base/src/algorithm/mesh/meshclipping.cpp [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+
+                    outIndices.push_back(*std::prev(start));
+                    std::copy(start, lineEnd, std::back_inserter(outIndices));
+                    outIndices.push_back(*lineEnd);
+                }
+                start = lineEnd;
+            }
+
+        } else {
+            throw Exception("Cannot clip, need line connectivity Strip or None",
```
Comment (martinfalk) [inline modules/base/src/algorithm/mesh/meshclipping.cpp:427]: so a line loop won't work?

File: modules/base/src/algorithm/mesh/meshclipping.cpp [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+        [&interpolateFunctors](const std::vector<uint32_t>& indices,
+                               const std::vector<float>& weights, std::optional<vec3> position,
+                               std::optional<vec3> normal) -> uint32_t {
+        uint32_t res = 0;
+        for (auto& fun : interpolateFunctors) res = fun(indices, weights, position, normal);
+        return res;
+    };
+
+    if (!posBuffer) {
+        throw Exception("Unsupported mesh type, vec3 position buffer not found",
```
Comment (martinfalk) [inline modules/base/src/algorithm/mesh/meshclipping.cpp:544]: hm, that could be an issue. There are some 2D meshes around as well. Guess that could be a separate PR if needed.

File: modules/base/src/processors/volumeboundaryplanes.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace inviwo {
+
+// The Class Identifier has to be globally unique. Use a reverse DNS naming scheme
+const ProcessorInfo VolumeBoundaryPlanes::processorInfo_{
+    "org.inviwo.VolumeBoundaryPlanes",  // Class identifier
+    "Volume Boundary Planes",           // Display name
+    "Data Creation",                    // Category
+    CodeState::Stable,                  // Code state
+    Tags::None,                         // Tags
```
Comment (martinfalk) [inline modules/base/src/processors/volumeboundaryplanes.cpp]: no tags?

File: modules/base/tests/regression/meshplaneclipping/config.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+  "image_test" :  {
+    "differenceTolerance" : {
+       "Canvas.png"   : 0.005
```
Comment (martinfalk) [inline modules/base/tests/regression/meshplaneclipping/config.json:4]: quite some tolerance...

File: modules/base/tests/unittests/meshcutting-test.cpp [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    ASSERT_EQ(loops.size(), 1);
+    ASSERT_EQ(loops[0].size(), 3);
+}
+
+TEST(MeshCutting, ClipMeshAgainstPlane) {
+
+    const auto mesh = meshutil::cube(mat4{1}, vec4{1, 1, 0, 1});
+    const Plane plane{vec3{0.5, 0.5, 0.5}, vec3{0, 1, 0}};
+
+    const auto clipped = meshutil::clipMeshAgainstPlane(*mesh, plane, true);
```
Comment (martinfalk) [inline modules/base/tests/unittests/meshcutting-test.cpp]: how about clipping a mesh against a plane where the entire mesh is on either side of the plane, i.e. no clipping happening?

**Review by martinfalk — APPROVED** [review]
nice!

File: modules/base/src/algorithm/mesh/meshclipping.cpp (hunk 8 above)
Comment (petersteneteg) [reply modules/base/src/algorithm/mesh/meshclipping.cpp:544]: clipping in 2D is a very different problem....

File: modules/base/tests/unittests/meshcutting-test.cpp (hunk 11 above)
Comment (petersteneteg) [reply modules/base/tests/unittests/meshcutting-test.cpp]: guess one could optimize that path if needed

File: modules/base/src/algorithm/mesh/meshclipping.cpp (hunk 7 above)
Comment (petersteneteg) [reply modules/base/src/algorithm/mesh/meshclipping.cpp:427]: No
